File: Advent_of_code_2025_Python/day3/day3.py
"""
Author: Matthieu Borelle
Description: Solution to Day 3 of the Advent of Code 2025 challenge.
"""

import numpy as np
import math
import re
from pathlib import Path


SCRIPT_DIR = Path(__file__).parent

def load_data(file_name):
    """Load data from a text file into a NumPy array."""
    with open(SCRIPT_DIR / file_name) as f:
        lines = [line.strip() for line in f]
    # Split each line into a list of integers
    list_of_digit_lists = [[int(digit) for digit in line] for line in lines]    
    return list_of_digit_lists  



## Part 1
def part1_solution(data_file_name):
    """
    Returns the solution of the part1 of the day3 challenge as an integer

    The total output joltage
    """
    #print(load_data(data_file_name))
    data = load_data(data_file_name)

    #print(data)    

    total_output_joltage = 0

    for i in range(0,len(data)): # for each line
        
        line_digits = data[i]
        
        max_list = [k for k, x in enumerate(line_digits[:-1]) if x == max(line_digits[:-1])] # remove the last digits for finding the max

        #print(max_list)

        list_candidate = [int(str(line_digits[k])+str(max(line_digits[k+1:]))) for k in max_list] 
        #[line_digits[k:k+1] for k in max_list]

        #print(list_candidate)

        max_output_joltage = max(list_candidate)

        total_output_joltage += max_output_joltage
   

    return total_output_joltage


## Part 2

# 003000000011
# 479103497921

# def look_for_max_digit_sequence(line_digits,sequence_length,candidate):
#     """ recursive search in the list from left to right """

#     #candidate = candidate
#     # legnth _left

#     if  len(line_digits)<=sequence_length:
#             candidate = line_digits
#             print(candidate)

#     elif len(line_digits)>sequence_length:
#         max_list = [k for k, x in enumerate(line_digits[:-sequence_length+1]) if x == max(line_digits[:-sequence_length+1])] 

#         print(max_list)

#         for k in max_list:
            
#             #line_digits[k+1:]
#             #sequence_length-1
#             #candidate
#             #candidate = int(str(line_digits[k])+str(look_for_max_digit_sequence(line_digits[k+1:],sequence_length-1,candidate)))
#             candidate = str(look_for_max_digit_sequence(line_digits[k+1:],sequence_length-1,candidate)) 
#             #print(candidate)
#     else:
#         print("Warning error")

#         # for k in max_list:

#     return candidate




def part2_solution(data_file_name):
    """
    Returns the solution of the part2 of the day3 challenge as an integer

    The total output joltage (with 12 digits in each bank's joltage output)

    """
    #print(load_data(data_file_name))
    data = load_data(data_file_name)

    #print(data)    

    total_output_joltage = 0

    number_digits_joltage = 12 #  > 0
    

    for i in range(0,len(data)): # for each line
        
        line_digits = data[i]
        remaining_number_digits = number_digits_joltage

        dynamic_line_digits = line_digits[:len(line_digits)-remaining_number_digits+1]
        #max_list = [k for k, x in enumerate(dynamic_line_digits) if x == max(dynamic_line_digits)]
        max_index = dynamic_line_digits.index(max(dynamic_line_digits))
   
        candidate = line_digits[max_index]

        last_max_index  = max_index
        remaining_number_digits -= 1


        while (remaining_number_digits>0):


            dynamic_line_digits =  line_digits[last_max_index+1:len(line_digits)-remaining_number_digits+1]

            #max_list = [k for k, x in enumerate(dynamic_line_digits) if x == max(dynamic_line_digits)] # remove the last digits for finding the max
            max_index = dynamic_line_digits.index(max(dynamic_line_digits))

            candidate = 10*candidate + dynamic_line_digits[max_index]

            last_max_index +=  max_index + 1


            remaining_number_digits -= 1    

            # if remaining_number_digits == len(line_digits) - max_index:

            #     remaining_number_digits = 0
            #     candidate = int(str(candidate) + str(line_digits[max_index+1:]))


        max_output_joltage = candidate

        total_output_joltage += max_output_joltage
   

    return total_output_joltage
    



if __name__ == "__main__": 

    print("Part 1 test:", part1_solution("day3_input_test.txt"))   # 357
    #print("Part 1 test:", part1_solution("day3_input_test2.txt")) # 
    print("Part 1 solution:", part1_solution("day3_input.txt")) #  17074
    print("---")
    print("Part 2 test:", part2_solution("day3_input_test.txt"))  # 3121910778619
    print("Part 2 solution:", part2_solution("day3_input.txt")) # 169512729575727

File: Advent_of_code_2025_C++/day3/day3.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <climits>

using namespace std;

// Function to load data from a text file
vector<vector<int>> load_data(const string& file_name) {
    ifstream file(file_name);
    vector<vector<int>> list_of_digit_lists;
    string line;
    while (getline(file, line)) {
        vector<int> digits;
        for (char digit_char : line) {
            digits.push_back(digit_char - '0'); // Convert char to int
        }
        list_of_digit_lists.push_back(digits);
    }
    return list_of_digit_lists;
}

// Part 1 solution
long long part1_solution(const string& data_file_name) {
    vector<vector<int>> data = load_data(data_file_name);
    long long total_output_joltage = 0;

    for (size_t i = 0; i < data.size(); ++i) {
        const vector<int>& line_digits = data[i];

        // Find indices of the max digit (excluding the last digit)
        vector<size_t> max_list;
        int max_digit = *max_element(line_digits.begin(), line_digits.end() - 1);
        for (size_t k = 0; k < line_digits.size() - 1; ++k) {
            if (line_digits[k] == max_digit) {
                max_list.push_back(k);
            }
        }

        // Generate candidates and find the maximum
        vector<long long> list_candidate;
        for (size_t k : max_list) {
            int next_max = *max_element(line_digits.begin() + k + 1, line_digits.end());
            long long candidate = line_digits[k] * 10 + next_max;
            list_candidate.push_back(candidate);
        }

        long long max_output_joltage = *max_element(list_candidate.begin(), list_candidate.end());
        total_output_joltage += max_output_joltage;
    }

    return total_output_joltage;
}

// Part 2 solution
long long part2_solution(const string& data_file_name) {
    vector<vector<int>> data = load_data(data_file_name);
    long long total_output_joltage = 0;
    const int number_digits_joltage = 12;

    for (size_t i = 0; i < data.size(); ++i) {
        const vector<int>& line_digits = data[i];
        int remaining_number_digits = number_digits_joltage;
        vector<int> dynamic_line_digits(line_digits.begin(), line_digits.end() - remaining_number_digits + 1);

        // Find the first max digit
        auto max_it = max_element(dynamic_line_digits.begin(), dynamic_line_digits.end());
        size_t max_index = distance(dynamic_line_digits.begin(), max_it);
        long long candidate = dynamic_line_digits[max_index];
        size_t last_max_index = max_index;
        remaining_number_digits--;

        // Iteratively find the next max digits
        while (remaining_number_digits > 0) {
            dynamic_line_digits.assign(
                line_digits.begin() + last_max_index + 1,
                line_digits.end() - remaining_number_digits + 1
            );
            max_it = max_element(dynamic_line_digits.begin(), dynamic_line_digits.end());
            max_index = distance(dynamic_line_digits.begin(), max_it);
            candidate = candidate * 10 + dynamic_line_digits[max_index];
            last_max_index += max_index + 1;
            remaining_number_digits--;
        }

        total_output_joltage += candidate;
    }

    return total_output_joltage;
}

int main() {
    cout << "Part 1 test: " << part1_solution("day3_input_test.txt") << endl;   // 357
    cout << "Part 1 solution: " << part1_solution("day3_input.txt") << endl;   // 17074
    cout << "---" << endl;
    cout << "Part 2 test: " << part2_solution("day3_input_test.txt") << endl;  // 3121910778619
    cout << "Part 2 solution: " << part2_solution("day3_input.txt") << endl; // 169512729575727
    return 0;
}

